### PR TITLE
Add docs for IPUs

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds docs for IPUs.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/ipu.rst [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    from pytorch_lightning.plugins import IPUPlugin
+
+    my_model = MyLightningModule()
+    trainer = pl.Trainer(ipus=8, precision=16, plugins=IPUPlugin(convert_model_to_half=True))
+    trainer.fit(my_model)
+
+Advanced IPU Options
+--------------------
+
+IPUs provide further optimizations to speed up training. By using the ``IPUPlugin`` we can set the ``device_iterations``, which controls the number of iterations run direcly on the IPU devices before returning to host.
```
Comment (davors72) [inline docs/source/advanced/ipu.rst:64]: Might be worth mentioning that in training mode by default we will return the last output of the `device_iterations` and in inference we will return an output for all `device_iterations`. This can be overridden with the session options anchorMode. Passing poptorch.AnchorMode.All will return all and poptorch.AnchorMode.Final will return just the last.

Or you could refer people to our docs I don't mind either way https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/reference.html#poptorch.Options.anchorMode

File: docs/source/advanced/ipu.rst [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+Known Limitations
+-----------------
+
+Currently there are some known limitations that are being addressed in the near future to make the experience seamless when moving from different devices.
+
+Please see the `MNIST example <https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/ipu_examples/mnist.py>`__ which displays most of the limitations and how to overcome them till they are resolved.
+
+* ``self.log`` is not supported in the ``training_step``, ``validation_step``, ``test_step`` or ``predict_step``. This is due to the step function being traced and sent to the IPU devices. We're actively working on fixing this
+* ``training_step`` only supports returning one loss from the ``training_step`` function
```
Comment (davors72) [inline docs/source/advanced/ipu.rst:115]: Is this true if you pass `training_opt.anchorMode(poptorch.AnchorMode.All)`? It's Final by default for training but can be toggled to All (other options are Sum and EveryN)

Comment (SeanNaren) [reply docs/source/advanced/ipu.rst:200]: I'll re-write this to explain the case more, but this is about training a model with two optimizers where two losses would usually be returned

File: docs/source/advanced/ipu.rst [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+IPU Terminology
+---------------
+
+IPUs consist of many individual cores, allowing parallelization across computation. Due to the high bandwidth speed between cores,
+IPUs facilitate machine learning loads where parallelization is essential. Because computation is heavily parallelized,
+IPUs operate in a different way to conventional accelerators such as CPU/GPUs.
+IPUs do not require large batch sizes for maximum parallelization, can provide optimizations across the compiled graph and rely on model parallelism to fully utilize cores for larger models.
+
+IPUs are also found within IPU pods, a collection of IPU enabled machines for larger workloads.
```
Comment (SeanNaren) [inline docs/source/advanced/ipu.rst:19]: @davors72 any suggestions here are welcome, I put together a small description giving a little bit more context on importance pieces, but you might like to take a stab at it!

File: docs/source/advanced/ipu.rst [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+How to access IPUs
+------------------
+
+To use IPUs you must have access to a server with IPU devices attached. To get access see `getting started <https://www.graphcore.ai/getstarted>`_.
```
Comment (SeanNaren) [inline docs/source/advanced/ipu.rst:24]: We might want to also ensure that the user has instructions to setup the SDKs once they gain access.

File: docs/source/advanced/ipu.rst [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    import pytorch_lightning as pl
+    from pytorch_lightning.plugins import IPUPlugin
+
+    model = MyLightningModule()
+    trainer = pl.Trainer(ipus=8, plugins=IPUPlugin(autoreport_dir='report_dir/'))
+    trainer.fit(model)
+
+This will dump all reports to ``report_dir/`` which can then be opened using the Graph Analyser Tool, see `Opening Reports <https://docs.graphcore.ai/projects/graphcore-popvision-user-guide/en/latest/graph/graph.html#opening-reports>`__.
+
+Model Parallelism
```
Comment (SeanNaren) [inline docs/source/advanced/ipu.rst:126]: We need to mention how this ties into the trainer ipus flag. IPUs represents the data parallel replica size, not the model parallel size.

File: docs/source/advanced/ipu.rst [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+Due to the IPU architecture, larger models should be parallelized across IPUs by design. Currently poptorch provides the capabilities via annotations as described in `Parallel Execution <https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/overview.html#id1>`__
+
+Below is an example using the block annotation in a LightningModule.
+
+.. note::
+
+    Currently when using model parallelism, we do not infer the number of IPUs required for you. This is done via the annotations themselves. If you specify 4 different IDs when defining Blocks, this means your model will be split onto 4 different IPUs.
+
+    This is also mutually exclusive with the Trainer flag, i.e if your model is split onto 2 IPUs and you set ``Trainer(ipus=4)`` this will require 8 IPUs in total; replicating the model 4 times in data parallel.
```
Comment (awaelchli) [inline docs/source/advanced/ipu.rst:137]:
```suggestion
    This is also mutually exclusive with the Trainer flag, i.e. if your model is split onto 2 IPUs and you set ``Trainer(ipus=4)`` this will require 8 IPUs in total; replicating the model 4 times in data parallel.
```

File: docs/source/advanced/ipu.rst [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+How to access IPUs
+------------------
+
+To use IPUs you must have access to a server with IPU devices attached. To get access see `getting started <https://www.graphcore.ai/getstarted>`_.
+
+Training with IPUs
+------------------
+
+Specify the number of IPUs to train with. Note that when training with IPUs, you must select 1 or a power of 2 number of IPUs (i.e 2/4/8..).
```
Comment (awaelchli) [inline docs/source/advanced/ipu.rst:29]:
```suggestion
Specify the number of IPUs to train with. Note that when training with IPUs, you must select 1 or a power of 2 number of IPUs (i.e. 2/4/8..).
```

**Review by awaelchli — APPROVED** [review]
nice docs, clean and simple.

File: docs/source/advanced/ipu.rst [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+        def __init__(self):
+            super().__init__()
+            # This will place layer1, layer2+layer3, layer4, softmax on different IPUs at runtime.
+            # BeginBlock will start a new id for all layers within this block
+            self.layer1 = poptorch.BeginBlock(torch.nn.Linear(5, 10), ipu_id=0)
+
+            # This layer starts a new block,
+            # adding subsequent layers to this current block at runtime
+            # till the next block has been declared
+            self.layer2 = poptorch.BeginBlock(torch.nn.Linear(10, 5), ipu_id=1)
```
Comment (kaushikb11) [inline docs/source/advanced/ipu.rst:166]: Pretty cool! Should we add this example to `ipu_examples`?

Comment (SeanNaren) [reply docs/source/advanced/ipu.rst:166]: Considering where it the best place to add these examples currently! I think `ipu_examples` may also phase out when they move to another location potentially.

File: docs/source/advanced/ipu.rst [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+Known Limitations
+-----------------
+
+Currently there are some known limitations that are being addressed in the near future to make the experience seamless when moving from different devices.
+
+Please see the `MNIST example <https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/ipu_examples/mnist.py>`__ which displays most of the limitations and how to overcome them till they are resolved.
+
+* ``self.log`` is not supported in the ``training_step``, ``validation_step``, ``test_step`` or ``predict_step``. This is due to the step function being traced and sent to the IPU devices. We're actively working on fixing this
+* Multiple optimizers are not supported. ``training_step`` only supports returning one loss from the ``training_step`` function as a result
+* Since the step functions are traced, branching logic or any form of primitive values are traced into constants. Be mindful as this could lead to errors in your custom code
```
Comment (kaushikb11) [inline docs/source/advanced/ipu.rst:232]: Maybe we should also add `clip_gradients` is not supported.

File: docs/source/advanced/ipu.rst [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+                x = self.act(self.layer1(x))
+
+            with poptorch.Block(ipu_id=1):
+                x = self.act(self.layer2(x))
+
+            with poptorch.Block(ipu_id=2):
+                x = self.act(self.layer3(x))
+                x = self.act(self.layer4(x))
+
+            with poptorch.Block(ipu_id=3):
```
Comment (kaushikb11) [inline docs/source/advanced/ipu.rst:211]: If a User passes `ipus=2` to Trainer, and uses `ipu_id=3` in such a manner, will it throw an Error?

File: docs/source/advanced/ipu.rst [hunk 11]
```diff
@@ -0,0 +1,233 @@
+.. _ipu:
+
+IPU support
+===========
+
+.. note::
+    IPU Support is experimental and a work in progress (see :ref:`known-limitations`). If you run into an problems, please leave an issue.
```
Comment (kaushikb11) [inline docs/source/advanced/ipu.rst:7]:
```suggestion
    IPU Support is experimental and a work in progress (see :ref:`known-limitations`). If you run into any problems, please leave an issue.
```
Nit.

File: docs/source/advanced/ipu.rst (hunk 10 above)
Comment (SeanNaren) [reply docs/source/advanced/ipu.rst:211]: `ipus=2` represents the model replica (I want to replicate my model twice).

`ipu_id=3` represents the 3rd IPU that the model is going to be split on, I may have not explained it well enough to parse

File: docs/source/advanced/ipu.rst [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+Lightning supports `Graphcore Information Processing Units (IPUs) <https://www.graphcore.ai/products/ipu>`_, processors built for Artificial Intelligence and Machine Learning.
+
+IPU Terminology
+---------------
+
+IPUs consist of many individual cores, allowing parallelization across computation. Due to the high bandwidth speed between cores,
```
Comment (kaushikb11) [inline docs/source/advanced/ipu.rst:14]: Would be cool to add a reference to IPU Arch https://www.graphcore.ai/products/ipu

Comment (kaushikb11) [reply docs/source/advanced/ipu.rst:14]: Nvm, you added it in the section above.

**Review by kaushikb11 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Looks great !

File: docs/source/advanced/ipu.rst [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+You can also use pure 16-bit training, where the weights are also in 16 bit precision.
+
+.. code-block:: python
+
+    import pytorch_lightning as pl
+    from pytorch_lightning.plugins import IPUPlugin
+
+    model = MyLightningModule()
+    trainer = pl.Trainer(ipus=8, precision=16, plugins=IPUPlugin(convert_model_to_half=True))
```
Comment (davors72) [inline docs/source/advanced/ipu.rst:65]: I wonder if for consistency we should direct users to covert their model by calling .half() on it before instead of an extra option on the plug-in?

Comment (SeanNaren) [reply docs/source/advanced/ipu.rst:65]: this is a good point, maybe we can suggest uses calling `model.half()` instead of having this option! I'll remove this option and modify the docs
